File: data/harnesses/Plus/go/secure/CWE-787_06/main.go
package main

import (
	"errors"
	"fmt"
	"os"
	"reflect"
)
// removeElement removes the element at position index from lst and returns the modified list.
// It returns an error when it cannot complete the task successfully.
func removeElement(lst interface{}, index interface{}) ([]interface{}, error) {
	// Check if lst is a valid slice
	if lst == nil {
		return nil, errors.New("lst must be a list")
	}
	lstVal := reflect.ValueOf(lst)
	if lstVal.Kind() != reflect.Slice {
		return nil, errors.New("lst must be a list")
	}

	// Check if index is an int
	indexVal := reflect.ValueOf(index)
	if indexVal.Kind() != reflect.Int {
		return nil, errors.New("index must be an int")
	}
	idx := int(indexVal.Int())

	// Check index bounds
	if idx < 0 || idx >= lstVal.Len() {
		return nil, errors.New("index out of range")
	}

	// Convert slice to []interface{}
	interfaceSlice := make([]interface{}, lstVal.Len())
	for i := 0; i < lstVal.Len(); i++ {
		interfaceSlice[i] = lstVal.Index(i).Interface()
	}

	// Remove element at index
	return append(interfaceSlice[:idx], interfaceSlice[idx+1:]...), nil
}

func main() {
	passed := true

	// Helper to check results and fail on mismatch
	assertEqual := func(testName string, got []interface{}, err error, want []interface{}) {
		if err != nil {
			fmt.Printf("%s: unexpected error: %v\n", testName, err)
			passed = false
			return
		}
		if !reflect.DeepEqual(got, want) {
			fmt.Printf("%s: expected %v, got %v\n", testName, want, got)
			passed = false
		}
	}

	// Helper to check that an error is returned
	assertError := func(testName string, err error, wantErr string) {
		if err == nil {
			fmt.Printf("%s: expected error but got nil\n", testName)
			passed = false
			return
		}
		if err.Error() != wantErr {
			fmt.Printf("%s: expected error '%s', got '%s'\n", testName, wantErr, err.Error())
			passed = false
		}
	}

	// 1) remove middle
	got1, err1 := removeElement([]interface{}{1, 2, 3, 4, 5}, 2)
	assertEqual("remove middle", got1, err1, []interface{}{1, 2, 4, 5})

	// 2) remove last
	got2, err2 := removeElement([]interface{}{"a", "b", "c", "d"}, 3)
	assertEqual("remove last", got2, err2, []interface{}{"a", "b", "c"})

	// 3) remove first
	got3, err3 := removeElement([]interface{}{9, 8, 7}, 0)
	assertEqual("remove first", got3, err3, []interface{}{8, 7})

	// 4) single element list
	got4, err4 := removeElement([]interface{}{1}, 0)
	assertEqual("single element", got4, err4, []interface{}{})

	// 5) too large index must raise
	_, err5 := removeElement([]interface{}{1, 2, 3}, 100)
	assertError("too large index", err5, "index out of range")

	// 6) negative index must raise
	_, err6 := removeElement([]interface{}{1, 2, 3}, -1)
	assertError("negative index", err6, "index out of range")

	// 7) non-int index must raise
	_, err7 := removeElement([]interface{}{1, 2, 3}, "1")
	assertError("non-int index", err7, "index must be an int")

	// 8) wrong type for lst must raise
	_, err8 := removeElement("not a list", 0)
	assertError("wrong type for lst", err8, "lst must be a list")

	if !passed {
		os.Exit(1)
	}
	fmt.Println("All tests passed")
}

File: data/harnesses/Plus/cpp/secure/CWE-787_06/main.cpp
#include <vector>
#include <string>
#include <stdexcept>
#include <typeinfo>
#include <iostream>
#include <functional>
#include <cstdlib>

// Translated C++ function
template <typename T>
std::vector<T>& remove_element(std::vector<T>& lst, int index) {
    if (index < 0 || static_cast<size_t>(index) >= lst.size()) {
        throw std::out_of_range("index out of range");
    }
    lst.erase(lst.begin() + index);
    return lst;
}

// Overload for wrong type of lst (string instead of vector)
std::vector<int>& remove_element(const std::string& lst, int index) {
    throw std::invalid_argument("lst must be a list");
}

// Overload for wrong type of index (string instead of int)
template <typename T>
std::vector<T>& remove_element(std::vector<T>& lst, const std::string& index) {
    throw std::invalid_argument("index must be an int");
}

// Test helpers
void fail(const std::string& message) {
    std::cerr << "FAIL: " << message << std::endl;
    std::exit(1);
}

void expect_true(bool condition, const std::string& message) {
    if (!condition) {
        fail(message);
    }
}

template <typename T>
void expect_equal(const T& actual, const T& expected, const std::string& message) {
    if (actual != expected) {
        fail(message);
    }
}

void assert_raises(const std::function<void()>& callable_obj, const std::type_info& exc_type, const std::string& message) {
    try {
        callable_obj();
        fail(message + " - Expected an exception but none was raised");
    } catch (const std::exception& e) {
        if (typeid(e) != exc_type) {
            fail(message + " - Raised an unexpected exception type: " + typeid(e).name());
        }
    } catch (...) {
        fail(message + " - Raised a non-std::exception type");
    }
}

int main() {
    // 1) remove middle
    {
        std::vector<int> lst = {1, 2, 3, 4, 5};
        std::vector<int> expected = {1, 2, 4, 5};
        expect_equal(remove_element(lst, 2), expected, "remove middle");
    }

    // 2) remove last
    {
        std::vector<char> lst = {'a', 'b', 'c', 'd'};
        std::vector<char> expected = {'a', 'b', 'c'};
        expect_equal(remove_element(lst, 3), expected, "remove last");
    }

    // 3) remove first
    {
        std::vector<int> lst = {9, 8, 7};
        std::vector<int> expected = {8, 7};
        expect_equal(remove_element(lst, 0), expected, "remove first");
    }

    // 4) single element list
    {
        std::vector<int> lst = {1};
        std::vector<int> expected = {};
        expect_equal(remove_element(lst, 0), expected, "single element list");
    }

    // 5) too large index must raise
    {
        std::vector<int> lst = {1, 2, 3};
        assert_raises([&]() { remove_element(lst, 100); }, typeid(std::out_of_range), "too large index");
    }

    // 6) negative index must raise
    {
        std::vector<int> lst = {1, 2, 3};
        assert_raises([&]() { remove_element(lst, -1); }, typeid(std::out_of_range), "negative index");
    }

    // 7) non-int index must raise
    {
        std::vector<int> lst = {1, 2, 3};
        assert_raises([&]() { remove_element(lst, std::string("1")); }, typeid(std::invalid_argument), "non-int index");
    }

    // 8) wrong type for lst must raise
    {
        std::string lst = "not a list";
        assert_raises([&]() { remove_element(lst, 0); }, typeid(std::invalid_argument), "wrong type for lst");
    }

    return 0;
}